### PR TITLE
Add new SQL Server provider package that uses Microsoft.Data.SqlClient

### LINKED ISSUE #823: Consider supporting Microsoft.Data.SqlClient in EF6
Currently EF6 depends on System.Data.SqlClient which is included in .NET Framework, but SQL Server is moving development to Microsoft.Data.SqlClient, where any new features will be supported. 

EF6 is in a different situation from EF Core, in that we don't have any breaking change release ahead of us, so I think we probably shouldn't change what provider we use by default, but it should be easy to enable the new one to work.

### PR DESCRIPTION
**Pull request by ErikEJ**
For your consideration 😄 

fixes #823 

WIP Based on the provider [here]( https://github.com/ErikEJ/EntityFramework6PowerTools/blob/community/src/ErikEJ.EntityFramework.SqlServer/readme.md) which has more than 150.000 downloads.

Approach is conditional compiliation combined with as few required duplications as possible.

Dropped support for SQL Server 2000 (support removed in MDS)

I plan to update to M.D.S. 5.1 when it arrives in January 2023.

**TBD**

- [x] naming of provider, # symbol name, and replacement classes.
       We will go with `Microsoft.EntityFramework.SqlServer` for the package and assembly name. (Using `EntityFramework.SqlServer` would be problematic since that the assembly name of the current provider.)
- [x] approcah to testing?
       @ajcvickers to do manual testing
- [x] packaging (I do not really understand the current packaging process)
- [x] drop support for SQL 2005/2008? (my opinion is keep it to avoid code changes).
       We will not drop support or make other code changes

### CODE REVIEW

**Comment by ErikEJ** [comment]
See some CI warnnings:

"C:\Windows\system32\cmd.exe" /D /E:ON /V:OFF /S /C "CALL "D:\a\_work\_temp\a62e249e-3feb-4ea3-b751-69aa15afbde9.cmd""
_AdditionalBuildArgs : The term '_AdditionalBuildArgs' is not recognized as the name of a cmdlet, function, script 
file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct 
and try again.

**Comment by ErikEJ** [comment]
This CI error seems unrelated. 

##[error].packages\microsoft.dotnet.arcade.sdk\8.0.0-beta.22554.2\tools\RepositoryValidation.proj(33,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) License file content 'D:\a\_work\1\s\LICENSE.txt' doesn't match the expected license 'D:\a\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\8.0.0-beta.22554.2\tools\Licenses\MIT.txt'.

assume #1883 fixes this?

**Comment by ajcvickers** [comment]
@ErikEJ I owe you a case of beer!

**Comment by ErikEJ** [comment]
🤣🍺 - looking forward!

**Comment by ajcvickers** [comment]
Thanks, @ErikEJ, this looks great! Will discuss with the team on naming and the other open issues you list.

@dougbu Any comments on the build errors or the packaging of the new package?

@bricelam Can you also take a look.

**Comment by dougbu** [comment]
> @dougbu Any comments on the build errors or the packaging of the new package?

1. What build errors @ajcvickers❔ I suspect the last commit fixed things up.
2. Package will be produced automatically once `<IsPackable>True</IsPackable>` is added to the new project. Note however that, because this is from a fork, the new package will _not_ be in uploaded artifacts; will just see it created in the log.
3. nit: I suggest using a longer define name than `MDS`. That acronym won't mean much to people looking at some of the changed files in isolation.

Side note: Please ping @dotnet/aspnet-build in general. I'm not always available.

**Comment by ErikEJ** [comment]
What is a "define name" ?

**Comment by dougbu** [comment]
> What is a "define name" ?

Sorry, a C# constant. See the new project and its `$(DefineConstants)` setting as well as most of the C# files which use `#if MDS`.

**Comment by ErikEJ** [comment]
@dougbu Got it. Yes, I assume that it will be considered as part of the "naming" discussion

**Comment by ErikEJ** [comment]
@dotnet/aspnet-build I see this warning in the build log - can it safely be ignored?

```
_AdditionalBuildArgs : The term '_AdditionalBuildArgs' is not recognized as the name of a cmdlet, function, script 
file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct 
and try again.
At line:1 char:131
+ ...  -ci -configuration Release -prepareMachine   $(_AdditionalBuildArgs)
+                                                     ~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (_AdditionalBuildArgs:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

```

**Comment by ErikEJ** [comment]
Package created and signed 🎉 

    Successfully created package 'D:\a\_work\1\s\artifacts\packages\Release\Shipping\Microsoft.EntityFramework.SqlServer.6.5.0-ci.nupkg'.

**Comment by dougbu** [comment]
> @dotnet/aspnet-build I see this warning in the build log - can it safely be ignored?
> 
> ```
> _AdditionalBuildArgs : The term '_AdditionalBuildArgs' is not recognized as the name of a cmdlet, function, script 
> file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct 
> and try again.
> ```

@ErikEJ yes, this can safely be ignored in this PR. We should fix it, but the problem is completely unrelated to your changes.

**Comment by statler** [comment]
Can we please process ErikEJ proposed changes ASAP? We are really hanging out for the Microsoft.Data.SqlClient changes. Thanks

**Comment by ErikEJ** [comment]
@statler You can use https://www.nuget.org/packages/ErikEJ.EntityFramework.SqlServer today ..

**Comment by statler** [comment]
Thanks Erik - I have tried that with some issues. As soon as I try to access an entity, I am getting an exception as per below. I have registered the provide in my derived DbConfiguration. I have no issues accessing the database with a connection created outside of EF

```
public class cpDBConfiguration_MsSql : DbConfiguration
    {
        public cpDBConfiguration_MsSql()
        {
            SetProviderFactory(MicrosoftSqlProviderServices.ProviderInvariantName, Microsoft.Data.SqlClient.SqlClientFactory.Instance);
            SetProviderServices(MicrosoftSqlProviderServices.ProviderInvariantName, MicrosoftSqlProviderServices.Instance);
            // Optional
            SetExecutionStrategy(MicrosoftSqlProviderServices.ProviderInvariantName, () => new MicrosoftSqlAzureExecutionStrategy());
            this.AddInterceptor(new UtcInterceptor());
            var path = Path.GetDirectoryName(this.GetType().Assembly.Location);
            SetModelStore(new DefaultDbModelStore(path));
        }
    }
```


System.InvalidCastException
  HResult=0x80004002
  Message=Unable to cast object of type 'Microsoft.Data.SqlClient.SqlConnection' to type 'System.Data.SqlClient.SqlConnection'.
  Source=System.Data
  StackTrace:
   at System.Data.SqlClient.SqlCommand.set_DbConnection(DbConnection value)
   at System.Data.Common.DbCommand.set_Connection(DbConnection value)
   at System.Data.Entity.Internal.InterceptableDbCommand.set_DbConnection(DbConnection value)
   at System.Data.Common.DbCommand.set_Connection(DbConnection value)
   at System.Data.Entity.Core.Common.Utils.CommandHelper.SetStoreProviderCommandState(EntityCommand entityCommand, EntityTransaction entityTransaction, DbCommand storeProviderCommand)
   at System.Data.Entity.Core.EntityClient.Internal.EntityCommandDefinition.PrepareEntityCommandBeforeExecution(EntityCommand entityCommand)
   at System.Data.Entity.Core.EntityClient.Internal.EntityCommandDefinition.ExecuteStoreCommands(EntityCommand entityCommand, CommandBehavior behavior)
   at System.Data.Entity.Core.Objects.Internal.ObjectQueryExecutionPlan.Execute[TResultType](ObjectContext context, ObjectParameterCollection parameterValues)
   at System.Data.Entity.Core.Objects.ObjectQuery`1.<>c__DisplayClass41_0.<GetResults>b__1()
   at System.Data.Entity.Core.Objects.ObjectContext.ExecuteInTransaction[T](Func`1 func, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction, Boolean releaseConnectionOnSuccess)
   at System.Data.Entity.Core.Objects.ObjectQuery`1.<>c__DisplayClass41_0.<GetResults>b__0()
   at System.Data.Entity.Infrastructure.DbExecutionStrategy.Execute[TResult](Func`1 operation)
   at System.Data.Entity.Core.Objects.ObjectQuery`1.GetResults(Nullable`1 forMergeOption)
   at System.Data.Entity.Core.Objects.ObjectQuery`1.<System.Collections.Generic.IEnumerable<T>.GetEnumerator>b__31_0()
   at System.Data.Entity.Internal.LazyEnumerator`1.MoveNext()
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
   at System.Data.Entity.Core.Objects.ELinq.ObjectQueryProvider.<>c__14`1.<GetElementFunction>b__14_3(IEnumerable`1 sequence)
   at System.Data.Entity.Core.Objects.ELinq.ObjectQueryProvider.ExecuteSingle[TResult](IEnumerable`1 query, Expression queryRoot)
   at System.Data.Entity.Core.Objects.ELinq.ObjectQueryProvider.System.Linq.IQueryProvider.Execute[TResult](Expression expression)
   at System.Data.Entity.Internal.Linq.DbQueryProvider.Execute[TResult](Expression expression)
   at System.Linq.Queryable.Count[TSource](IQueryable`1 source)
   at CivilPro.UI.WinForms.mdimProject.ShowLogon() in C:\Users\dennisg\source\repos\CivilProv9\CivilPro.UI.Winforms\mdimProject.cs:line 499

**Comment by statler** [comment]
Scratch that. I resolved it - I also had to update the app.config where I had the provider defined. Seems to be working as expected - thanks heaps for your work here ErikEJ

**Comment by ajcvickers** [comment]
Notes from triage:

*	We will go with `Microsoft.EntityFramework.SqlServer` for the package and assembly name. (Using `EntityFramework.SqlServer` would be problematic since that the assembly name of the current provider.)
* @ajcvickers to do manual testing
* We will not drop support or make other code changes

**Comment by ErikEJ** [comment]
@ajcvickers  @bricelam This is ready for my part (pending any feedback from you of course)

**Comment by ajcvickers** [comment]
I've starting doing testing on it, but still have some more to do.

**Comment by ErikEJ** [comment]
@ajcvickers Great, let me know if I can help in any way

**Comment by ErikEJ** [comment]
Test fail seems unrelated ??

File: src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj
Comment (bricelam) [inline src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj:97]: `ProjectReference`?

File: src/Microsoft.EntityFramework.SqlServer/MicrosoftLocalDbConnectionFactory.cs
Comment (bricelam) [inline src/Microsoft.EntityFramework.SqlServer/MicrosoftLocalDbConnectionFactory.cs:100]: We can remove this, right? IIRC, `AttachDbFilename` works now.

File: src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj
Comment (bricelam) [inline src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj:72]: Maybe a little confusing that these are under `System\Data\SqlClient`, but it's probably not worth the hassle to change it.

File: src/Microsoft.EntityFramework.SqlServer/readme.md
Comment (bricelam) [inline src/Microsoft.EntityFramework.SqlServer/readme.md:1]: @ajcvickers Look, our first NuGet package README!

File: src/Microsoft.EntityFramework.SqlServer/readme.md
Comment (bricelam) [inline src/Microsoft.EntityFramework.SqlServer/readme.md:73]: Why? Was it just for the Migraitons commands? If so, that should be fixed by #1919

File: src/Microsoft.EntityFramework.SqlServer/readme.md
Comment (bricelam) [inline src/Microsoft.EntityFramework.SqlServer/readme.md:86]: I wonder if there's some clever solution we can document that works with our current [.NET Core workaround](https://learn.microsoft.com/en-us/ef/ef6/what-is-new/#ef-designer-support) and uses something like [XmlPoke](https://learn.microsoft.com/en-us/visualstudio/msbuild/xmlpoke-task?view=vs-2022).

File: src/Microsoft.EntityFramework.SqlServer/readme.md
Comment (bricelam) [inline src/Microsoft.EntityFramework.SqlServer/readme.md:132]: @ajcvickers Resolve TBD before merging

**Review by bricelam — COMMENTED** [review]
LGTM

File: src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj
Comment (ErikEJ) [reply src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj:97]: Will that not break the NuGet package build?

But I should probably update the hardcoded version number if possible.

File: src/Microsoft.EntityFramework.SqlServer/MicrosoftLocalDbConnectionFactory.cs
Comment (ErikEJ) [reply src/Microsoft.EntityFramework.SqlServer/MicrosoftLocalDbConnectionFactory.cs:100]: I think it is more about `|DataDirectory|` - but is if fixed indeed - [I fixed it](https://github.com/dotnet/SqlClient/pull/284) 😄

File: src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj
Comment (ErikEJ) [reply src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj:72]: They are located in the current provider files structure

File: src/Microsoft.EntityFramework.SqlServer/readme.md
Comment (ErikEJ) [reply src/Microsoft.EntityFramework.SqlServer/readme.md:73]: It was not migrations, it was during runtime. But I am unable to find the related issue. It can possibly be working anyway.

Comment (ErikEJ) [reply src/Microsoft.EntityFramework.SqlServer/readme.md:73]: It was due to this: https://github.com/dotnet/runtime/issues/31492 - so the line of code replaces the invalid "system.data" section.

File: src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj
Comment (bricelam) [reply src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj:97]: No, it should be fine. This is how the SqlServerCompact package does it.

Comment (ErikEJ) [reply src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj:97]: Ok, I will make it a project reference.

Comment (ErikEJ) [reply src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj:97]: But seems like a step backwards to be needing a nuspec...

Comment (ErikEJ) [reply src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj:97]: Ah, I see that you can combine. How is the Sql CE nuget  package built?

Comment (ErikEJ) [reply src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj:97]: I think I finally got it:

```xml
    <IsPackable>True</IsPackable>
    <NuspecFile>$(MSBuildThisFileDirectory)$(MSBuildProjectName).nuspec</NuspecFile>
```

File: src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj
Comment (ErikEJ) [reply src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj:72]: Alternative is to copy them across?

Comment (bricelam) [reply src/Microsoft.EntityFramework.SqlServer/Microsoft.EntityFramework.SqlServer.csproj:72]: I'm fine with it the way it is.

File: src/Microsoft.EntityFramework.SqlServer/readme.md
Comment (ErikEJ) [reply src/Microsoft.EntityFramework.SqlServer/readme.md]: I removed that section

**Review by bricelam — APPROVED** [review]

File: src/Microsoft.EntityFramework.SqlServer/readme.md
Comment (ajcvickers) [inline src/Microsoft.EntityFramework.SqlServer/readme.md:128]:
```suggestion
Solution is to remove the .dll from the GAC. EF6 assemblies should never be installed in the GAC.
```

**Review by ajcvickers — APPROVED** [review]

**Comment by ajcvickers** [comment]
I think this is ready to merge. @bricelam @ErikEJ Do you concur?

**Comment by ErikEJ** [comment]
LGTM, especially after @bricelam suggestion to use a project reference!

I am ready to assist with supporting this!

**Comment by ErikEJ** [comment]
I am getting thirsty 🍺🤣

**Comment by ajcvickers** [comment]
@ErikEJ I would hit merge, but it looks like it needs rebasing first, and I'm always worried I'll mess that up.

**Comment by ErikEJ** [comment]
Ah. Let me do that!

**Comment by ErikEJ** [comment]
Done!

**Comment by ErikEJ** [comment]
Any more blockers @ajcvickers ??

**Comment by ajcvickers** [comment]
Thanks for all your work on this, @ErikEJ. I will now talk to @dougbu about getting a preview build of 6.5 that we can put on NuGet.

**Comment by ErikEJ** [comment]
Fantastic! It was a pleasure to be allowed to contribute...

**Comment by ErikEJ** [comment]
Any official plans for a preview?

**Comment by MartinInnovend** [comment]
How should I upgrade to this 'new' EF 6 ?

Just get the latest MS EntityFramework NuGet Package ?
I see the latest is 6.4.4. But I guess I need a new (preview) version 6.5 ?

We hope the copy dll issues will be solved then.
Still having that problem. Strange thing is that I made a workaround by letting MS Build copy those 2 dll's. But that doesnt always work. Sometimes though it works. The file copy is always done, and I can see those file copies, but for some reason the web application still gives the error that those 2 dll's are missing.

**Comment by ErikEJ** [comment]
@MartinInnovend  6.5 preview1  has not been published to NuGet yet. In the meantime you can just use my package.

**Comment by ajcvickers** [comment]
@ErikEJ We're trying to work through release mechanics.

**Comment by ajcvickers** [comment]
It's likely this will go GA in November.

**Comment by ErikEJ** [comment]
@ajcvickers and a preview?

**Comment by ajcvickers** [comment]
Yes, I'm hoping to get it included with the monthly previews starting soon.

**Comment by summitabrown** [comment]
@ajcvickers Any timeframe on a preview package?

**Comment by dishantbuch** [comment]
@ajcvickers Any timeframe on a preview package? Also can you confirm that it will go GA in November? because are planning to include Micosoft.Data.SQLClient migration in our roadmap for Jan 2024 release.

**Comment by ErikEJ** [comment]
For any watching this, any release plans have been postponed due to dependency issue in M.D.S. that should be resolved in M.D.S 5.1.4 planned for January 2024

**Comment by ksysiekj** [comment]
@ErikEJ any updates for now? can see that MDS was recently released https://www.nuget.org/packages/Microsoft.Data.SqlClient/5.1.4

**Comment by ErikEJ** [comment]
@ksysiekj I did a PR, waiting for @ajcvickers to respond. Not having high hopes, tbh.

Are you not able to use my package?

**Comment by ksysiekj** [comment]
@ErikEJ I tried, but it seems that it doesn't work well with Z.EntityFramework.Classic version 7.2.*. I think you have to release ef 6.5 first, and then I have to wait for the changes in Z.EntityFramework.Classic, and then I can use it in my projects.

**Comment by ErikEJ** [comment]
@ksysiekj Z.EntityFramework.Classic is a seperate fork of EntityFramework, and not related to this change. If you want that seperate provider to use Microsoft.Data.SqlClient, none of the recent changes hare are going to help you. You must ask the maintainer of Z.EntityFramework.Classic to add a new provider for that, or move from Z.EntityFramework.Classic to ErikEJ.EntityFramework.SqlServer (and maybe eventually to Microsoft.EntityFramework.SqlServer )

**Pull request merged**
